src/manybody_deepmd.F

```diff
@@ -0,0 +1,315 @@
+                                     deepmd_data_type
+   USE kinds, ONLY: dp
+   USE pair_potential_types, ONLY: pair_potential_pp_type, &
+                                   pair_potential_single_type, &
+                                   deepmd_pot_type, &
+                                   deepmd_type
+   USE particle_types, ONLY: particle_type
+   USE physcon, ONLY: angstrom, &
+                      evolt
+#ifdef __DEEPMD
+   USE deepmd_wrapper, ONLY: nnp, &
+                             create_nnp, &
+                             delete_nnp_c, &
+                             compute_nnp
+#endif
+
+#include "./base/base_uses.f90"
+
+   IMPLICIT NONE
+
+   PRIVATE
+   LOGICAL, SAVE, PRIVATE :: do_create(2) = .TRUE.
+   INTEGER, SAVE, PRIVATE :: do_count = 0
+   PUBLIC deepmd_energy_store_force_virial, deepmd_add_force_virial
+
+   CHARACTER(len=*), PARAMETER, PRIVATE :: moduleN = 'manybody_deepmd'
+
+CONTAINS
+
+! **************************************************************************************************
+!> \brief ...
+!> \param particle_set ...
+!> \param cell ...
+!> \param atomic_kind_set ...
+!> \param potparm ...
+!> \param fist_nonbond_env ...
+!> \param pot_deepmd ...
+!> \param para_env ...
+! **************************************************************************************************
+   SUBROUTINE deepmd_energy_store_force_virial(particle_set, cell, atomic_kind_set, potparm, fist_nonbond_env, &
+                                               pot_deepmd, para_env)
+      TYPE(particle_type), POINTER             :: particle_set(:)
+      TYPE(cell_type), POINTER                 :: cell
+      TYPE(atomic_kind_type), POINTER          :: atomic_kind_set(:)
+      TYPE(pair_potential_pp_type), POINTER    :: potparm
+      TYPE(fist_nonbond_env_type), POINTER     :: fist_nonbond_env
+      REAL(kind=dp), TARGET                            :: pot_deepmd
+      TYPE(cp_para_env_type), OPTIONAL, &
+         POINTER                                :: para_env
+
+      CHARACTER(LEN=*), PARAMETER :: &
+         routineN = 'deepmd_energy_store_force_virial', &
+         routineP = moduleN//':'//routineN
+
+#ifdef __DEEPMD
+      INTEGER                                  :: i, iat, iat_use, ikind, &
+                                                  jkind, n_atoms, n_atoms_use, &
+                                                  output_unit
+      LOGICAL                                  :: do_parallel
+      LOGICAL, ALLOCATABLE                     :: use_atom(:)
+      REAL(kind=dp)                            :: lattice(3, 3)
+      TYPE(pair_potential_single_type), &
+         POINTER                                :: pot
+      TYPE(deepmd_data_type), POINTER            :: deepmd_data
+      TYPE(deepmd_pot_type), POINTER             :: deepmd
+      TYPE(nnp), SAVE                                :: deepmd_tmp_pot1, &
+                                                        deepmd_tmp_pot2
```

Comment:
No. If you have to preserve data you have to add it to a suitable env. No global vars or shared state.